### PR TITLE
fix(drizzle): make mutation statements resilient to connection drops via transaction-wrapped retry

### LINKED ISSUE #911: bug: @agentuity/drizzle resilient proxy retries non-idempotent INSERT queries, causing duplicate rows
### Summary

`@agentuity/drizzle`'s `createResilientSQLProxy` wraps all `unsafe()` calls with `executeWithRetry()`, including INSERT statements. When an INSERT with a server-generated primary key (e.g., `uuid().defaultRandom()`) succeeds on Postgres but the connection drops before the response reaches the client, the retry re-executes the INSERT — creating a duplicate row with a different auto-generated UUID.

### Root Cause

In `src/postgres.ts`, the resilient SQL proxy intercepts `unsafe()` and wraps it with retry logic:

```typescript
if (prop === 'unsafe') {
    return (query, params) => {
        const makeExecutor = (useValues) => client.executeWithRetry(async () => {
            const currentRaw = client.raw;
            const q = currentRaw.unsafe(query, params);
            return useValues ? q.values() : q;
        });
        const result = makeExecutor(false);
        return Object.assign(result, {
            values: () => makeExecutor(true),
        });
    };
}
```

`executeWithRetry` (in `@agentuity/postgres`) retries up to 3 times on retryable connection errors (`ECONNRESET`, `connection closed`, `EPIPE`, etc.). This is correct for idempotent operations (SELECT, UPDATE, DELETE) but causes data duplication for INSERT statements where the primary key is server-generated.

**The sequence:**

1. Drizzle calls `db.insert(table).values({ /* no explicit id */ }).returning()`
2. Drizzle generates SQL: `INSERT INTO table (...) VALUES (...) RETURNING *`
3. Resilient proxy intercepts `unsafe(sql, params)` → wraps with `executeWithRetry`
4. **Attempt 1:** INSERT executes on Postgres → row created with UUID `aaaa` → connection hiccup before response arrives
5. `executeWithRetry` catches retryable error (e.g., `ECONNRESET`) → retries
6. **Attempt 2:** Same INSERT executes again → Postgres generates a NEW UUID `bbbb` → row created → response arrives
7. Client receives row `bbbb`, unaware that row `aaaa` also exists
8. Row `aaaa` is orphaned in the database

### How to Reproduce

**Environment:** Remote Postgres over SSL (e.g., `ion-usc.agentuity.cloud`). The issue is more likely with higher-latency or less stable connections.

**Schema:**
```typescript
import { pgTable, uuid, text, timestamp } from '@agentuity/drizzle';

const items = pgTable('items', {
    id: uuid('id').primaryKey().defaultRandom(),
    name: text('name'),
    status: text('status').default('pending'),
    createdAt: timestamp('created_at', { withTimezone: true }).defaultNow(),
});
```

**Code that triggers the bug:**
```typescript
import { createPostgresDrizzle } from '@agentuity/drizzle';
import * as schema from './schema';

const { db } = createPostgresDrizzle({
    connectionString: 'postgresql://...@remote-host:5432/mydb?sslmode=require',
    schema,
});

// This INSERT is NOT idempotent — each execution creates a new row
const [item] = await db
    .insert(schema.items)
    .values({ name: 'test', status: 'creating' })
    .returning();

// If executeWithRetry retried, there are now 2 rows in the DB
// but `item` only references the second one
```

**Evidence from production debugging:**

We added file-based logging around the insert call:
```
[BEFORE INSERT] 2026-02-10T21:03:27.482Z    ← called ONCE
[AFTER INSERT]  2026-02-10T21:03:27.650Z id=89323b64...  ← returned ONE id
```

But the database contained TWO rows:
```
89323b64... | active   | 2026-02-10 21:03:27.60777    ← returned to client
0204395b... | creating | 2026-02-10 21:03:27.607241   ← orphaned duplicate
```

Timestamps are 0.5ms apart — consistent with an immediate retry after a transient connection error.

### Impact

- Orphaned rows accumulate in the database (status stuck in initial state, no background processing)
- Users see duplicate entries in the UI (one working, one broken)
- Background processing (e.g., sandbox creation) only runs for the row returned to the caller — the orphaned row never transitions

### Suggested Fixes

**Option A: Detect and skip retry for INSERT statements**
Parse the SQL string — if it starts with `INSERT`, don't retry (or retry with a warning).

**Option B: Document the idempotency requirement**
Add guidance that users should generate primary keys client-side when using tables with auto-generated UUIDs:
```typescript
import { randomUUID } from 'node:crypto';
const id = randomUUID();
await db.insert(table).values({ id, ... }).onConflictDoNothing().returning();
```

**Option C: Use savepoints for automatic retry safety**
Wrap each retried operation in a savepoint so the duplicate INSERT would fail within the same transaction context.

**Option D: Add opt-out mechanism**
Allow callers to disable retry for specific operations:
```typescript
// Hypothetical API
db.insert(table).values({...}).returning().$noRetry();
```

### Current Workaround

Generate UUID client-side and use `onConflictDoNothing()`:
```typescript
import { randomUUID } from 'node:crypto';

const id = randomUUID();
const rows = await db
    .insert(table)
    .values({ id, name: 'test', status: 'creating' })
    .onConflictDoNothing()
    .returning();

let item = rows[0];
if (!item) {
    // Retry caused no-op insert — fetch the existing row
    const [existing] = await db.select().from(table).where(eq(table.id, id)).limit(1);
    item = existing;
}
```

### Versions

- `@agentuity/drizzle`: latest (installed via agentuity project)
- `@agentuity/postgres`: latest
- Runtime: Bun
- Database: PostgreSQL on `ion-usc.agentuity.cloud` (remote, SSL)

### PR DESCRIPTION
**Pull request by jhaynie**
## Summary

- Make INSERT, UPDATE, and DELETE statements resilient to PostgreSQL connection drops (e.g., during Ion proxy rollouts) by wrapping them in `BEGIN/COMMIT` transactions with automatic retry
- Use a lazy thenable pattern to prevent double execution when Drizzle calls `.values()` on mutation results
- Only SELECT queries continue with bare retry (safe since they're idempotent)

## Problem

During Ion proxy rollouts, PostgreSQL connections are dropped, causing `ERR_POSTGRES_CONNECTION_CLOSED` errors. The SDK's resilient SQL proxy retried SELECT queries but explicitly skipped INSERT/UPDATE/DELETE to avoid:
- Duplicate rows from retried INSERTs (server-generated IDs)
- Double-applied changes from retried UPDATEs (`counter = counter + 1`)
- Repeated side effects from retried DELETEs (cascade triggers)

This meant all mutations failed hard during connection drops with no recovery.

## Solution

Wrap mutation statements in transactions before retrying:

```
BEGIN → INSERT/UPDATE/DELETE → COMMIT
```

**Why this is safe:** PostgreSQL guarantees that uncommitted transactions are automatically rolled back when the connection drops. If the connection fails at any point before `COMMIT` completes on the server, no changes are applied, and the retry starts a fresh transaction on the new connection.

**Known limitation:** If the connection drops after the server processes `COMMIT` but before the client receives the response (~<1ms window), the changes are committed and a retry would re-apply them. This is an inherent limitation of any retry-based approach and can be mitigated with unique constraints / `ON CONFLICT` at the application level.

## Changes

| File | Change |
|------|--------|
| `packages/drizzle/src/postgres.ts` | Renamed `isNonRetryableInsert` → `isMutationStatement`; detects INSERT, UPDATE, DELETE; wraps in BEGIN/COMMIT with lazy thenable for single-execution guarantee |
| `packages/drizzle/test/proxy.test.ts` | Updated 15 tests, added 4 new tests (rollback, reconnection, lazy thenable double-call prevention, direct await) |

## Query Behavior Matrix

| Query Type | Before | After |
|------------|--------|-------|
| SELECT | Bare retry | Bare retry (unchanged) |
| INSERT | ❌ No retry | ✅ Transaction-wrapped retry |
| UPDATE | ⚠️ Bare retry (unsafe for non-idempotent) | ✅ Transaction-wrapped retry |
| DELETE | ⚠️ Bare retry (unsafe for cascades) | ✅ Transaction-wrapped retry |

## Testing

- ✅ 95 tests pass, 0 fail
- ✅ Build clean
- ✅ Typecheck clean
- ✅ Reviewed for safety and edge cases

Fixes #911

## Summary by CodeRabbit

* **Bug Fixes**
  * Detection now treats INSERT/UPDATE/DELETE (including mutations inside CTEs) as requiring transactional retries; scanning correctly ignores strings/comments and nested parentheses. Transactional execution ensures a single mutation per query with rollback on error and safer retry behavior for commit-observation edge cases.

* **Tests**
  * Expanded tests to verify transaction-wrapped execution, rollback sequencing, concurrency/thenable/.values() behaviors, reconnection/retry paths, and numerous whitespace/comment/CTE edge cases.

### CODE REVIEW

File: packages/drizzle/test/proxy.test.ts [hunk 1]
```diff
@@ -341,22 +544,258 @@ describe('createResilientSQLProxy', () => {
...
+		const mockUnsafe = mock((query: string, _params?: unknown[]) => {
+			unsafeCalls.push(query);
+			callIndex++;
+			// First call: BEGIN succeeds
+			// Second call: INSERT fails
+			// Third call: ROLLBACK succeeds
+			if (callIndex === 2) {
+				return Promise.reject(new Error('query failed'));
+			}
+			const result = Promise.resolve([{ id: 1 }]);
```
[the remark this answers is not in the document]
Comment (jhaynie) [reply packages/drizzle/test/proxy.test.ts]: we don't need StructuredError for test cases or simple error handling
